### DISCUSSION TITLE
Hallucinations - how to recognize and solve?

### DISCUSSION THREAD

**Discussion by nikola1975** [openai/whisper#1369] [Q&A]
Hello. I am using Whisper API for some time now. It works very good for big languages and almost acceptable for small ones.

However, occasionally it hallucinates and as part of the transcription, it sends back repeated words or phrases. Sometimes, this can be one word repeated many times, other times it is few words one after the other and then repeated again (like a repeated phrase).

I am trying to recognize this pattern and in that case I would send the audio back to transcription again, as it usually does not repeat for the same audio.

I have tried two approaches:
1. Send the text to OpenAI GPT 3.5 API to recognize suspicious transcription. As I am doing it as part of the other action, this would not bring additional cost, but unfortunately, it is not able to report back consistently. I instructed it to return “[SUSPICIOUS]” if it believes there could be mistakes, but it does so for perfectly good transcriptions as well.
2. Recognize repeated words with Python function. If one word is repeated, it has worked ok, but the problem is with phrases. I need to cut the text and then recognize 2 repeated words, 3 repeated words, 4 repeated words. I am not even sure this will work. Is there more intelligent way than that?

Thanks,
Nikola

**Comment by phineas-pta** [comment]
it is possible to pre-process audios before sending to api?

**Reply by nikola1975** [reply]
I am doing audio normalization before sending it to the API, if that is what you meant?

**Reply by phineas-pta** [reply]
that isn't enough, what i mean is removing silence / no-speech segments, the main source of hallucination

**Reply by nikola1975** [reply]
Aha, I understand. I will try that as well. What I can do is remove the silence, but the music or background noise is proving more challenging.

Also, the examples I noticed did not have either silence or non-speech parts. They are more frequent in "smaller" languages, and if the quality is not the highest, I would say that is why I wanted to check the transcriptions and re-transcribe if they seem suspicious.

**Reply by phineas-pta** [reply]
hallucinations on speech segments 🤔 now that's worrysome, the solution for low-resource languages i think is fine-tuning; because detecting repeated string within a text isn't trivial as search for 1-2-3 words, it involves advanced text-mining

**Reply by nikola1975** [reply]
I am now trying to do something like this. It works for some examples, ie. stops part of the problems and returns the audio to re-transcription. Not sure this is the best approach. 

```
def check_for_repetition(text, chunk_size=100, repetition_threshold=14, immediate_repeat_threshold=6):
    words = re.findall(r'\b\w+\b', text)  # only consider words, ignore punctuation

    chunks = [words[i:i + chunk_size] for i in range(0, len(words), chunk_size)]
    for chunk in chunks:
        word_frequencies = Counter(chunk)
        
        for word, frequency in word_frequencies.items():
            if frequency > repetition_threshold:
                logging.warning(f'Repetition detected for word "{word}" {frequency} times in chunk "{" ".join(chunk)}"')
                return True  # repetition detected

        # check for immediate repetition
        repeat_counter = 1
        for i in range(1, len(chunk)):
            if chunk[i] == chunk[i-1]:
                repeat_counter += 1  # increase counter if word is same as previous
                if repeat_counter >= immediate_repeat_threshold:
                    logging.warning(f'Immediate repetition detected for word "{chunk[i]}" {repeat_counter} times in chunk "{" ".join(chunk)}"')
                    return True  # immediate repetition detected
            else:
                repeat_counter = 1  # reset counter if word is different

    return False
```

**Reply by phineas-pta** [reply]
2 arguments `repetition_threshold` and `immediate_repeat_threshold` can have lots of values depending on audios, you can't find any "good enough" default value for all cases (too high = false negative, too low = false positive)

<details>
<summary>illustrating examples</summary>
<i>1st case: false negative</i>
<pre>
example1 = """Right now, it's Giles Peterson.
What time is it?
3 o'clock.
I can't hear you!
The only thing that I'm saying here is that we are making a joyful noise to people that
don't suppose to be dying yet.
Music is the spirit.
Music is life.
Life is music.
This is Giles Peterson.
Music is life.
Music is life.
Music is life.
Music is life.
I'm not so much of a technical player myself, probably not that much of an intellectual
player as some other musicians.
What I do is express.
Music is life.
Music is life.
Music is life.
Music is life.
Music is life."""
</pre>
the programs checks and oops ... it passes!
<pre>
check_for_repetition(example1)
# False
</pre>
you only find out the hallucination when the transcript is done<br /><br />
<i>2nd case: false positive</i>
<pre>
example2 = """Come a little bit closer, hear what I have to say.
Just like children sleeping, we could dream this night away.
There's a full moon rising, let's go dancing in the light.
We know where the music's playing, let's go out and feel the night.
It's alright, it's alright."""
</pre>
you decide to lower <code>repetition_threshold</code>
<pre>
check_for_repetition(example2, repetition_threshold=5)
# WARNING:root:Repetition detected for word "s" 6 times in chunk "..."
# True
</pre>
you look at the warning + transcript and change the threshold again ...<br /><br />
also you find out the regex <code>r'\b\w+\b'</code> isn't immune to english  <code>'s</code> or french <code>n'</code> or many other languages frequent special punctuations ...
</details>

**TLDR** the function isn't reliable nor automatic: to find values for 2 thresholds, you'll have to actually look at the transcript at each usage, meaning the function is useless, because you write function to not look at the transcript; and also it doesn't cover all edge cases

**Reply by mrfragger** [reply]
I've experienced hallucinations quite a few times.  So if it is due to silence then does anyone know how I believe whisperX uses silero VAD voice activity detection to preprocess files before feeding it to whisper.  In other words are the silent parts just timestamped to be ignored by whisper (I hope this is the case) or is it making a new file with the cut out silence parts?  ffmpeg can automatically detect silent parts and trim them out.

something like so 

`for f in *.opus ; do ffmpeg -c:a libopus -vbr off -b:a 32k -ar 48000 -af highpass=500,lowpass=1000,afftdn,aformat=channel_layouts=stereo,volume=12dB, "silenceremove=start_periods=1:stop_periods=-1:start_threshold=-50dB:stop_threshold=-50dB:start_silence=1:start_duration=2:stop_duration=5:detection=peak",dynaudnorm "$f" ; done`

the stop_duration=5 any amount over 5 seconds to take out.  Normal conversation is 60dB so anything below 50dB is pretty safe to remove.
Green show original audio clip with silence parts. Black shows anything over over 30 seconds remove, Red shows anything over 15 seconds removed
 
![silence](https://github.com/openai/whisper/assets/80211672/68a00ae5-72bd-4fad-874d-d26f656c30da)

Here are a couple quotes from other discussions related to this.  So what's the max silence before it goes haywire? 

> Also, if you don't remove pauses longer than 29-30s (more or less equal than whisper internal chunks) it will hallucinate and write garbage or get stuck in a loop of garbage generation. They're working on it but for now the best thing is to cut the non speech parts with a VAD such as silero

and this one

> VAD filtering: Voice Activity Detection (VAD) from Pyannote.audio (https://huggingface.co/pyannote/voice-activity-detection) is used as a preprocessing step to remove reliance on whisper timestamps and only transcribe audio segments containing speech. add --vad_filter flag, increases timestamp accuracy and robustness (requires more GPU mem due to 30s inputs in wav2vec2)

On your part 2 detecting repeated words I'm pretty sure I could do that with sed. At least I believe so.

**Reply by phineas-pta** [reply]
idk much about VAD, but silero vad & pyannote are open source, you can actually look at source code instead of wondering

2ndly, it's called **voice** activity detection, not *silence* detection, that's how it's different from volume-based detection with `ffmpeg`, non-silence non-speech segments also cause hallucination

on detecting repeated words with `sed`, good luck tinkering with regex + repetition thresholds, you'll soon find edge cases all over the place with each new transcript

**Reply by nikola1975** [reply]
@phineas-pta Yes, you are right about false positives and false negatives. I have created this function to decrease number of problems. At this moment, with the lack of proper solution, I will try to catch most problems - it will still be better then nothing. That was also the reason of my posting here, to see if there are some better ideas.

If I catch the problem, I am trying to transcribe it again - and if it happens once more, I will let the transcription as it is.

**Reply by phineas-pta** [reply]
the solution for now is to pre-process audios with VAD

if you want to process subtitles post-transcription, you can try alignment with other speech-to-text models like `wav2vec2` as in https://github.com/m-bain/whisperX ; other not-NLP text processing not recommended to remove repetition

other interesting point is that when you re-transcribe, hallucinations disappear, maybe it's something special to openai api, because with local-pc (colab) whisper still hallucinates

**Reply by mrfragger** [reply]
> 
> on detecting repeated words with `sed`, good luck tinkering with regex + repetition thresholds, you'll soon find edge cases all over the place with each new transcript

yea you're correct it's almost impossible to filter repeating words like there are many many issues which would be legit, but at the end of the day it's not a transcript for a book per se so gotta just figure out what's worse.  100 instances of repeating words that shouldn't be or a few instances of repeating words that should but get taken out to a single case.  Anyway here's the [sed script](https://github.com/ggerganov/whisper.cpp/discussions/957) for pro-nouns for whisper when it goes berserk

**Reply by phineas-pta** [reply]
you should properly write a script with variables/functions etc. because a one-line 35k-character command may be helpful to some non-technical users who want something ready-to-use, but it's difficult and painful to other people who want to modify/adapt to their needs

also you should use something like gist.github.com to share code, don't share zip (because of security concerns)

**Comment by LaurinmyReha** [comment]
maybe this is useful for someone here still:
I wrote a article about this which might help you fix this yourself:
https://nyra-labs.com/research/killing-hallucinations
which improves this behaviour quite drastically. Feel free to try out the model too :)
https://huggingface.co/nyralabs/CrisperWhisper2.0_large
